### PR TITLE
test/ unit tests for create Impersonation Requests

### PR DESCRIPTION
**Pull request by Suvidh-kaushik**
Date: 8/06/2025

Developer Name: Suvidh Kaushik

---

## Issue Ticket Number

- Closes: #2200 

## Description

- This PR contains the unit tests for the model, validator and services of the create impersonation request logic
- **This PR must be merged after the Feature PR -  #2439** 
- **The import errors and build fail is because the feature PR has not been merged yet** 

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

N/A

</details>

## Test Coverage
<details>
<summary>Screenshots</summary>

**Model Unit Test Coverage**

Screenshot 2025-06-18 012618

Screenshot 2025-06-18 012626

**Validator Unit Test Coverage**

Screenshot 2025-06-18 012958

Screenshot 2025-06-11 025036

**Services Unit Test Coverage**

Screenshot 2025-06-11 025555

Screenshot 2025-06-18 013727

</details>

## Additional Details

Design Doc - [LINK](https://docs.google.com/document/d/1Qt0-knoFEWDl6Gdxr0W1xhPy87H-ixwj9Y55Q0bKMUo/edit?tab=t.0#heading=h.w1nbw3goorza)

PRD - [LINK](https://docs.google.com/document/d/1KI7zp81vevaIX3nDNgTsCHDY1iWmad-XFaVgEi3nH2Q/edit?tab=t.0)

### What change is being made?

Add unit tests for the createImpersonationRequest feature, including fixtures, middlewares, models, and services.

### Why are these changes being made?

These changes are being made to validate and ensure the integrity of the new createImpersonationRequest feature by testing its components extensively. The tests aim to verify correct behavior, handle edge cases, and identify potential issues before the feature is deployed. This approach enhances reliability and aids in maintaining code quality.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: test/unit/services/impersonationRequests.test.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  TASK_REQUEST_MESSAGES,
+} from "../../../constants/requests";
+import userDataFixture from "./../../fixtures/user/user";
+const userData = userDataFixture();
+import addUser from "../../utils/addUser";
+import cleanDb from "../../utils/cleanDb";
+import {
+  impersonationRequestsBodyData,
+} from "../../fixtures/impersonation-requests/impersonationRequests";
+import { CreateImpersonationRequestModelDto, ImpersonationRequest } from "../../../types/impersonationRequest";
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/services/impersonationRequests.test.ts:19]: it is present in the feature PR which will be merged before this

File: test/unit/models/impersonationRequests.test.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import cleanDb from "../../utils/cleanDb";
+import * as impersonationModel from "../../../models/impersonationRequests";
+import {
+  impersonationRequestsBodyData,
+} from "../../fixtures/impersonation-requests/impersonationRequests";
+import { REQUEST_STATE, ERROR_WHILE_CREATING_REQUEST} from "../../../constants/requests";
+import addUser from "../../utils/addUser";
+import userDataFixture from "./../../fixtures/user/user";
+import sinon from "sinon";
+import { CreateImpersonationRequestModelDto } from "../../../types/impersonationRequest";
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:11]: It will be merged in the feature PR before this PR is merged

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  describe("createImpersonationRequestValidator", function () {
+    it("should validate for a valid create impersonation request", async function () {
+      req = {
+        body: requestBody
+      };
+      await createImpersonationRequestValidator(req as any, res as any, nextSpy);
+      expect(nextSpy.calledOnce).to.be.true;
+    });
+
+    it("should not validate for an invalid impersonation request on missing impersonatedUserId", async function () {
```
Comment (MayankBansal12) [inline test/unit/middlewares/impersonationRequests.test.ts:34]:
```suggestion
    it("should invalidate for an invalid impersonation request on missing impersonatedUserId", async function () {
```

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import sinon from "sinon";
+import { createImpersonationRequestValidator } from "../../../middlewares/validators/impersonationRequests";
+import { CreateImpersonationRequestBody } from "../../../types/impersonationRequest";
+const { expect } = chai;
+
+describe("Impersonation Request Validators", function () {
+  let req: any;
+  let res: any;
+  const requestBody: CreateImpersonationRequestBody = {
+    impersonatedUserId: "randomId",
```
Comment (MayankBansal12) [inline test/unit/middlewares/impersonationRequests.test.ts:11]: can we rename it to `mockRequestBody` as it is being used as template in diff places in this file

File: test/unit/services/impersonationRequests.test.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  });
+
+  afterEach(async () => {
+    await cleanDb();
+    sinon.restore();
+  });
+
+  describe("createImpersonationRequestService", () => {
+    beforeEach(async () => {
+      requestBody = impersonationRequestsBodyData[0];
```
Comment (MayankBansal12) [inline test/unit/services/impersonationRequests.test.ts:33]: `impersonationRequestsBodyData` is being used at diff places, what does it contain and how is it diff from `requestBody` defined earlier

File: test/unit/services/impersonationRequests.test.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    it("should return NotFound error with USER_NOT_FOUND if userId does not exist", async () => {
+      try {
+        await impersonationService.createImpersonationRequestService({
+          userId: "randomIs",
+          createdBy: "randomName",
+          impersonatedUserId: "randomImpersonatedId",
+          reason: "He asked",
+        });
+      } catch (err: any) {
```
Comment (MayankBansal12) [inline test/unit/services/impersonationRequests.test.ts:49]: avoid using `any`

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import chai from "chai";
+import sinon from "sinon";
+import { createImpersonationRequestValidator } from "../../../middlewares/validators/impersonationRequests";
+import { CreateImpersonationRequestBody } from "../../../types/impersonationRequest";
+const { expect } = chai;
+
+describe("Impersonation Request Validators", function () {
+  let req: any;
+  let res: any;
```
Comment (MayankBansal12) [inline test/unit/middlewares/impersonationRequests.test.ts:9]: avoid using `any`

Comment (MayankBansal12) [reply test/unit/middlewares/impersonationRequests.test.ts:9]:
```
let req: Partial<Request>;
let res: Partial<Response>;
```

see if this works

File: test/unit/middlewares/impersonationRequests.test.ts
Comment (MayankBansal12) [inline test/unit/middlewares/impersonationRequests.test.ts:1]: replace `any` from all the places for req, res with the ones from express

**Review by MayankBansal12 — CHANGES_REQUESTED** [review]

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 3 above)
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:34]: doesn't this complexify the reading of the test name

File: test/unit/services/impersonationRequests.test.ts (hunk 5 above)
Comment (Suvidh-kaushik) [reply test/unit/services/impersonationRequests.test.ts:33]: yes, it is an array which contains some mock data for the request creation, you can see more details about this in the fixtures file

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 7 above)
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:9]: have updated this

Comment (MayankBansal12) [reply test/unit/middlewares/impersonationRequests.test.ts:16]: @Suvidh-kaushik this one still remaining

File: test/unit/services/impersonationRequests.test.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+describe("Tests Impersonation Requests Service", () => {
+  let testUserId: string;
+  let mockRequestBody: CreateImpersonationRequestModelDto;
+  let userDetail;
+  const userData = userDataFixture();
+
+  beforeEach(async () => {
+    await cleanDb();
+    testUserId = await addUser(userData[20]);
+    userDetail = userData[20];
```
Comment (pankajjs) [inline test/unit/services/impersonationRequests.test.ts:24]: no need to put this in beforeEach block

File: test/unit/services/impersonationRequests.test.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+  afterEach(async () => {
+    await cleanDb();
+    sinon.restore();
+  });
+
+
+ describe("createImpersonationRequestService", () => {
+    beforeEach(async () => {
+      mockRequestBody = impersonationRequestsBodyData[0];
```
Comment (pankajjs) [inline test/unit/services/impersonationRequests.test.ts:35]: same

File: test/unit/services/impersonationRequests.test.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        id: "123",
+        status: REQUEST_STATE.PENDING,
+        isImpersonationFinished: false,
+        createdBy: mockRequestBody.createdBy,
+        createdFor: mockRequestBody.createdFor,
+        userId: mockRequestBody.userId,
+        reason: "He asked",
+        impersonatedUserId: testUserId,
+        createdAt:Timestamp.now(),
+        updatedAt:Timestamp.now()
```
Comment (pankajjs) [inline test/unit/services/impersonationRequests.test.ts:69]: This can be used from fixture.

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+  beforeEach(function () {
+    res = {
+      boom: {
+        badRequest: sinon.spy(),
+      },
+    };
+    nextSpy = sinon.spy();
+  });
+
```
Comment (pankajjs) [inline test/unit/middlewares/impersonationRequests.test.ts:31]: Not clearing spies in afterEach block

File: test/unit/services/impersonationRequests.test.ts (hunk 8 above)
Comment (Suvidh-kaushik) [reply test/unit/services/impersonationRequests.test.ts]: removed the userDetail from the beforeEach block also keeping the testUserId since we will use it more than 2 tests

**Review by pankajjs — CHANGES_REQUESTED** [review]

File: test/unit/models/impersonationRequests.test.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      try {
+        await impersonationModel.createImpersonationRequest(mockRequestBody);
+      } catch (error) {
+        expect(error.message).to.include(REQUEST_ALREADY_PENDING);
+      }
+    });
+
+    it("should allow different super users to create requests for same user", async () => {
+      const request1 = await impersonationModel.createImpersonationRequest({ ...impersonationRequestsBodyData[0],createdBy: "user1" });
+      const request2 = await impersonationModel.createImpersonationRequest({ ...impersonationRequestsBodyData[0],createdBy: "user2", userId:"122" });
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:51]: this is because the first request is using the fixture body which has a userId already present , since we are using the same fixture i updated the createBy and the userId of the second request so that it can simulate another super user details

File: test/unit/models/impersonationRequests.test.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        });
+      } catch (error) {
+        expect(error.message).to.include(ERROR_WHILE_CREATING_REQUEST);
+      }
+    });
+
+    it("should throw forbidden error if an APPROVED request with isImpersonationFinished as false is present", async ()=>{
+      try {
+        await impersonationModel.createImpersonationRequest({...impersonationRequestsBodyData[0],status:REQUEST_STATE.APPROVED});
+        await impersonationModel.createImpersonationRequest(impersonationRequestsBodyData[0]);
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:74]: i did not explicitly specify it there but in the fixture we have isImpersonationFinished as false hence i did not need to rewrite it again

File: test/unit/services/impersonationRequests.test.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import { CreateImpersonationRequestModelDto } from "../../../types/impersonationRequest";
+import { Timestamp } from "firebase-admin/firestore";
+const userQuery = require("../../../models/users");
+
+describe("Tests Impersonation Requests Service", () => {
+  let mockRequestBody: CreateImpersonationRequestModelDto = impersonationRequestsBodyData[0];
+  const userData = userDataFixture();
+
+  afterEach(async () => {
+    await cleanDb();
```
Comment (pankajjs) [inline test/unit/services/impersonationRequests.test.ts:19]: no necessary

**Review by pankajjs — APPROVED** [review]
LGTM

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]
